### PR TITLE
AWS Plugin to ensure ecs container insights are enabled

### PR DESCRIPTION
**Pull request by abdullahaslam306**

### CODE REVIEW

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Container Insights Enabled',
+    category: 'ECS',
+    domain: 'Containers',
+    description:
+      'Ensure that ECS clusters have CloudWatch Container Insights feature enabled.',
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:9]:
```suggestion
    description: 'Ensure that ECS clusters have CloudWatch Container Insights feature enabled.',
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                    return rcb();
+                }
+
+                if (listClusters.data.length === 0) {
+                    helpers.addResult(results, 0, 'No ECS clusters present', region);
+                    return rcb();
+                }
+     
+                for (var c in listClusters.data) {
+                    var clsuterARN = listClusters.data[c];
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:48]:
```suggestion
                for (var clusterARN of listClusters.data) {
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                        for ( var index in describeCluster.data.clusters) {
+             
+                            const cluster = describeCluster.data.clusters[index];
+                            if (cluster.settings.length > 0){ 
+                                for (var item of cluster.settings ){ 
+                                    if (item.name === 'containerInsights' ){
+
+                                        if ( item.value === 'enabled'){
+
+                                            helpers.addResult(
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:87]: use a flag like inightsEnabled and set it inside loop and add result outside the loop. Just to make sure we don't generate multiple results for same resource

File: plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                        err: null,
+                        data: descData
+                    }
+                }
+            }
+        },
+    }
+};
+
+describe('ECSContainerInsightsEnabled', function () {
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js:26]: write 1 test per each result

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+                                        if ( item.value === 'enabled'){
+                                             containerInsightsEnabled = true
+                                             break;
+                                        }
+                                    }
+
+                                }
+                            }
+                             if(containerInsightsEnabled){
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:70]: indentation seems off

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+                                    helpers.addResult(results, 0,
+                                    'ECS cluster container Insights is enabled', region, clusterARN);
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:72]:
```suggestion
                                    'ECS cluster has container Insights enabled', region, clusterARN);
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                                }
+                                else {
+                                    helpers.addResult(results, 2,
+                                                'ECS cluster container insights  not enabled',region, clusterARN);
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:76]:
```suggestion
                                                'ECS cluster does not have container insights enabled',region, clusterARN);
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: exports.js [hunk 8]
```diff
@@ -583,6 +583,7 @@ module.exports = {
 
         'fraudDetectorDataEncrypted'    : require(__dirname + '/plugins/aws/frauddetector/fraudDetectorDataEncrypted.js'),
 
+        'ecsContainerInsightsEnabled'    : require(__dirname + '/plugins/aws/ecs/ecsContainerInsightsEnabled.js'),
```
Comment (alphadev4) [inline exports.js:586]:
```suggestion
        'ecsContainerInsightsEnabled'   : require(__dirname + '/plugins/aws/ecs/ecsContainerInsightsEnabled.js'),
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            if (!listClusters.data.length) {
+                helpers.addResult(results, 0, 'No ECS clusters present', region);
+                return rcb();
+            }
+
+            for (var clusterARN of listClusters.data) {
+                var describeCluster = helpers.addSource(cache, source,
+                    ['ecs', 'describeCluster', region, clusterARN ]);
```
Comment (alphadev4) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:38]:
```suggestion
                    ['ecs', 'describeCluster', region, clusterARN]);
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+                if (!describeCluster || describeCluster.err ||!describeCluster.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to describe ECS cluster: ' +helpers.addError(describeCluster), region, clusterARN);
+                    continue;
+                }
+
+                if (
+                    describeCluster.data.clusters &&
+                    describeCluster.data.clusters.length
+                ) {
```
Comment (alphadev4) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:49]:
```suggestion
                if (describeCluster.data.clusters && describeCluster.data.clusters.length) {   
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                if (!describeCluster || describeCluster.err ||!describeCluster.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to describe ECS cluster: ' +helpers.addError(describeCluster), region, clusterARN);
+                    continue;
+                }
+
+                if (describeCluster.data.clusters && describeCluster.data.clusters.length) {   
+                    for ( var index in describeCluster.data.clusters) {
+
+                        let containerInsightsEnabled = false;
```
Comment (alphadev4) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:49]:
```suggestion
                    for (var index in describeCluster.data.clusters) {
                        let containerInsightsEnabled = false;
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+                    helpers.addResult(results, 3,
+                        'Unable to describe ECS cluster: ' +helpers.addError(describeCluster), region, clusterARN);
+                    continue;
+                }
+
+                if (describeCluster.data.clusters && describeCluster.data.clusters.length) {   
+                    for (var index in describeCluster.data.clusters) {
+                        let containerInsightsEnabled = false;
+                        const cluster = describeCluster.data.clusters[index];
+                        if (cluster.settings.length > 0){
```
Comment (alphadev4) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:50]:
```suggestion
                        const cluster = describeCluster.data.clusters[index];
                        
                        if (cluster.settings.length > 0) { 
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                                if (item.name === 'containerInsights' ){
+
+                                    if (item.value === 'enabled'){
+                                        containerInsightsEnabled = true;
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                        if (containerInsightsEnabled){
```
Comment (alphadev4) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:62]:
```suggestion
                        if (containerInsightsEnabled) {
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+                {
+                  "clusters": [{
+                    "name": "mycluster",
+                    "arn": "arn:aws:ecs:us-east-1:012345678911:cluster/testCluster",
+                    "settings": [{"name": "containerInsights", value: "enabled"}]
+                  }]}
+                
+            );
+            ecs.run(cache, {}, callback);
+        });
```
Comment (alphadev4) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js:121]:
```suggestion
        });
        
```

**Review by alphadev4 — APPROVED** [review]

File: plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+                expect(results.length).to.equal(1)
+                expect(results[0].status).to.equal(0)
+                expect(results[0].message).to.include('No ECS clusters present')
+                done()
+            };
+
+            const cache = createCache(
+                [],
+                {}
+            );
```
Comment (alphadev4) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js:39]:
```suggestion
            const cache = createCache([], {});
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+            if (!listClusters.data.length) {
+                helpers.addResult(results, 0, 'No ECS clusters present', region);
+                return rcb();
+            }
+
+            for (var clusterARN of listClusters.data) {
+                var describeCluster = helpers.addSource(cache, source,
+                    ['ecs', 'describeCluster', region, clusterARN]);
+        
+                if (!describeCluster || describeCluster.err ||!describeCluster.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:40]:
```suggestion
                if (!describeCluster || describeCluster.err ||!describeCluster.data ||
                    describeCluster.data.clusters || describeCluster.data.clusters.length) {
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+                        if (containerInsightsEnabled) {
+                            helpers.addResult(results, 0,
+                                'ECS cluster has container Insights enabled', region, clusterARN);
+                        } else {
+                            helpers.addResult(results, 2,
+                                'ECS cluster does not have container insights enabled',region, clusterARN);
+                        }             
+                    }            
+                } 
+            }
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:71]:
```suggestion
                const cluster = describeCluster.data.clusters[0];
                let containerInsightsEnabled = (cluster.settings && cluster.settings.length) ? cluster.settings.find(item => item.name == 'containerInsights' && item.value == ''enabled) : false;

                if (containerInsightsEnabled) {
                    helpers.addResult(results, 0,
                        'ECS cluster has container Insights enabled', region, clusterARN);
                } else {
                    helpers.addResult(results, 2,
                         'ECS cluster does not have container insights enabled', region, clusterARN);
                }             
            }
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+                
+            );
+            ecs.run(cache, {}, callback);
+        });
+
+        it('should give passing result if ecs cluster has container insights is enabled', function (done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('ECS cluster has container Insights enabled');
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.spec.js:103]:
```suggestion
                expect(results[0].message).to.include('ECS cluster has container insights enabled');
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        
+                if (!describeCluster || describeCluster.err ||!describeCluster.data ||
+                    describeCluster.data.clusters || describeCluster.data.clusters.length) {
+                    helpers.addResult(results, 3,
+                        'Unable to describe ECS cluster: ' +helpers.addError(describeCluster), region, clusterARN);
+                    continue;
+                }
+
+                const cluster = describeCluster.data.clusters[0];
+                let containerInsightsEnabled = (cluster.settings && cluster.settings.length) ? cluster.settings.find(item => item.name == 'containerInsights' && item.value == ''enabled) : false;
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:48]:
```suggestion
                let containerInsightsEnabled = (cluster.settings && cluster.settings.length) ? cluster.settings.find(item => item.name == 'containerInsights' && item.value == 'enabled') : false;
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+                const cluster = describeCluster.data.clusters[0];
+                let containerInsightsEnabled = (cluster.settings && cluster.settings.length) ? cluster.settings.find(item => item.name == 'containerInsights' && item.value == 'enabled') : false;
+
+                if (containerInsightsEnabled) {
+                    helpers.addResult(results, 0,
+                        'ECS cluster has container insights enabled', region, clusterARN);
+                } else {
+                    helpers.addResult(results, 2,
+                         'ECS cluster does not have container insights enabled', region, clusterARN);
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:55]:
```suggestion
                        'ECS cluster does not have container insights enabled', region, clusterARN);
```

File: plugins/aws/ecs/ecsContainerInsightsEnabled.js [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+                helpers.addResult(results, 0, 'No ECS clusters present', region);
+                return rcb();
+            }
+
+            for (var clusterARN of listClusters.data) {
+                var describeCluster = helpers.addSource(cache, source,
+                    ['ecs', 'describeCluster', region, clusterARN]);
+        
+                if (!describeCluster || describeCluster.err ||!describeCluster.data ||
+                    describeCluster.data.clusters || describeCluster.data.clusters.length) {
```
Comment (AkhtarAmir) [inline plugins/aws/ecs/ecsContainerInsightsEnabled.js:41]:
```suggestion
                    !describeCluster.data.clusters || !describeCluster.data.clusters.length) {
```

**Pull request merged**